opencxl/cxl/component/cxl_cache_manager.py

```diff
@@ -42,11 +43,65 @@ def __init__(
         self._upstream_fifo: FifoPair
         self._cache_device_component = None
 
+        self.device_entries = {}  # maps CQID -> received future packets associated with CQID
+        self.device_entry_lock = Lock()  # locks the above mapping
+
         super().__init__(upstream_fifo, downstream_fifo, label)
 
     def set_memory_device_component(self, cache_device_component: CxlMemoryDeviceComponent):
         self._cache_device_component = cache_device_component
 
+    async def register_cqid_listener(
+        self,
+        cqid: int,
+        cb: Callable[[BasePacket], None],
+        _timeout: Optional[int] = None,
+        _one_use: bool = True,
+    ):
+        """
+        Registers a callback action upon reception of a H2DRSP/H2DREQ packet with matching CQID.
+        Used to emulate the "device entry" model in CXL.cache spec. The listener should accept
+        a single CxlBasePacket parameter, which will contain the H2D packet object.
+
+        If `_one_use` is True, the callback action will automatically unregister itself after
+        one call. Note that if `_one_use` is false, the caller will have to manually handle
+        cancelling by raising CancelledError, just as they would for a generic asyncio Task.
+
+        If the operation times out, the callback action unconditionally unregisters itself.
+        """
+        async with self.device_entry_lock:
+            # two threads accessing self.device_entries simultaneously
+            # has the potential to SERIOUSLY break the emulator
+            if cqid in self.device_entries:
+                # cancel any currently running listeners, if they exist
+                self.device_entries[cqid].cancel()
+
+        loop = get_running_loop()
+        fut_pckt = loop.create_future()
+
+        async def _callback(fut: Future):
+            while True:
+                if _timeout:
+                    try:
+                        async with timeout(_timeout):
+                            await fut
+                    except TimeoutError as err:
+                        # this request was apparently lost by the host
+                        # clear the cqid entry for reuse
+                        async with self.device_entry_lock:
+                            del self.device_entries[cqid]
+                        raise CancelledError from err  # intentionally cancel the current task
+                else:
+                    await fut
+                cb(fut.result())
+                if _one_use:
+                    async with self.device_entry_lock:
+                        del self.device_entries[cqid]
+                    break
+
+        self.device_entries[cqid] = fut_pckt
+        create_task(_callback(fut_pckt))
```

Comment:
This function takes `cb`, a callback. Then, there is `_callback()`, another callback. Please rename one or both.